File: src/dotnet/core/utilities/Result.cs
using System;
using System.Runtime.InteropServices;

namespace Hyperion
{
    // Maps to core/utilities/Result.hpp
    [HypClassBinding(Name="Result")]
    [StructLayout(LayoutKind.Explicit, Size = 8)]
    public struct Result
    {
        [FieldOffset(0)]
        private PimplPtr errorPtr;

        public bool IsValid
        {
            get
            {
                return errorPtr.ptr == IntPtr.Zero;
            }
        }
    }
}

File: src/core/utilities/Result.cpp
#include <core/utilities/Result.hpp>

namespace hyperion {
namespace utilities {

class NullError final : public IError
{
public:
    NullError() = default;
    virtual ~NullError() override = default;

    virtual const String& GetMessage() const override
    {
        static const String message = "<null>";

        return message;
    }

    virtual ANSIStringView GetFunctionName() const override
    {
        return ANSIStringView();
    }
};

HYP_API const IError& GetNullError()
{
    static NullError null_error;

    return null_error;
}

} // namespace utilities
} // namespace hyperion